File: app/src/main/java/cn/ifafu/ifafu/mvp/web/WebPresenter.java
package cn.ifafu.ifafu.mvp.web;

import android.webkit.CookieManager;

import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.data.exception.NoAuthException;
import cn.ifafu.ifafu.mvp.base.BaseZFPresenter;
import cn.ifafu.ifafu.util.RxUtils;
import cn.ifafu.ifafu.util.SPUtils;

class WebPresenter extends BaseZFPresenter<WebContract.View, WebContract.Model> implements WebContract.Presenter {

    WebPresenter(WebContract.View view) {
        super(view, new WebModel(view.getContext()));
    }

    @Override
    public void onCreate() {
        String title = mView.getActivity().getIntent().getStringExtra("title");
        String url = mView.getActivity().getIntent().getStringExtra("url");
        if (title != null && url != null) {
            mView.setTitle(title);
            mView.loadUrl(url);
        } else {
            loadZF();
        }
    }

    private void loadZF() {
        mCompDisposable.add(mModel.getMainHtml()
                .map(s -> {
                    if (s.contains("请登录")) {
                        throw new NoAuthException();
                    } else {
                        String url = mModel.getMainUrl();
                        setCookie(url, SPUtils.get(Constant.SP_COOKIE).getString("ASP.NET_SessionId"));
                        return url;
                    }
                })
                .compose(RxUtils.ioToMain())
                .subscribe(url -> {
                    mView.loadUrl(url);
                }, throwable -> {
                    mView.loadUrl(mModel.getMainUrl());
                    onError(throwable);
                })
        );
    }

    private void setCookie(String url, String cookie) {
        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.flush();
        cookieManager.setAcceptCookie(true);
        cookieManager.setCookie(url, cookie);
    }
}

File: app/src/main/java/cn/ifafu/ifafu/mvp/web/WebPresenter.kt
package cn.ifafu.ifafu.mvp.web

import android.webkit.CookieManager
import cn.ifafu.ifafu.R
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.base.ifafu.BaseZFPresenter
import cn.ifafu.ifafu.mvp.web.WebContract.Presenter
import cn.ifafu.ifafu.mvp.web.WebContract.View
import cn.ifafu.ifafu.util.RxUtils
import cn.ifafu.ifafu.util.SPUtils

internal class WebPresenter(view: View) : BaseZFPresenter<View, WebContract.Model>(view, WebModel(view.context)), Presenter {

    override fun onCreate() {

        val title: String? = mView.activity.intent.getStringExtra("title")
        val referer: String? = mView.activity.intent.getStringExtra("referer")
        val url: String? = mView.activity.intent.getStringExtra("url")

        if (url != null && title != null) {
            mView.setTitle(title)
            if (referer != null) {
                setCookie(url, referer)
            }
            mView.loadUrl(url)
        } else {
            mView.setTitle(resId = R.string.title_web_mode)
            loadUrl(mModel.getMainUrl(), SPUtils.get(Constant.SP_COOKIE).getString("ASP.NET_SessionId"))
        }
    }

    private fun loadUrl(url: String, cookie: String) {
        mCompDisposable.add(mModel.loadMainHtml()
                .map {
                    setCookie(url, cookie)
                    url
                }
                .compose(RxUtils.ioToMain())
                .doOnSubscribe { mView.showLoading() }
                .doFinally { mView.hideLoading() }
                .subscribe({ mView.loadUrl(url) }, { throwable: Throwable ->
                    mView.loadUrl(mModel.getMainUrl())
                    onError(throwable)
                })
        )
    }

    private fun setCookie(url: String, cookie: String?) {
        val cookieManager: CookieManager = CookieManager.getInstance()
        cookieManager.flush()
        cookieManager.setAcceptCookie(true)
        cookieManager.setCookie(url, cookie)
    }
}